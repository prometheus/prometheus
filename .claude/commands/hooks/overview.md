# Claude Code Hooks for claude-flow

## Purpose
Automatically coordinate, format, and learn from Claude Code operations using hooks with MCP tool integration.

## Available Hooks

### Pre-Operation Hooks
- **pre-edit**: Validate and assign agents before file modifications
- **pre-bash**: Check command safety and resource requirements  
- **pre-task**: Auto-spawn agents for complex tasks

### Post-Operation Hooks
- **post-edit**: Auto-format code, train neural patterns, update memory
- **post-bash**: Log execution and update metrics
- **post-search**: Cache results and improve search patterns

### MCP Integration Hooks
- **mcp-initialized**: Persist swarm configuration
- **agent-spawned**: Update agent roster and memory
- **task-orchestrated**: Monitor task progress through memory
- **neural-trained**: Save pattern improvements

### Memory Coordination Hooks
- **memory-write**: Triggered when agents write to coordination memory
- **memory-read**: Triggered when agents read from coordination memory
- **memory-sync**: Synchronize memory across swarm agents

### Session Hooks
- **notify**: Custom notifications with swarm status
- **session-end**: Generate summary and save state
- **session-restore**: Load previous session state

## Configuration
Hooks are configured in `.claude/settings.json`:

```json
{
  "hooks": {
    "PreToolUse": [
      {
        "matcher": "^(Write|Edit|MultiEdit)$",
        "hooks": [{
          "type": "command",
          "command": "npx claude-flow hook pre-edit --file '${tool.params.file_path}' --memory-key 'swarm/editor/current'"
        }]
      }
    ],
    "PostToolUse": [
      {
        "matcher": "^(Write|Edit|MultiEdit)$",
        "hooks": [{
          "type": "command",
          "command": "npx claude-flow hook post-edit --file '${tool.params.file_path}' --memory-key 'swarm/editor/complete'"
        }]
      }
    ]
  }
}
```

## MCP Tool Integration in Hooks

Hooks automatically trigger MCP tools for coordination:

```javascript
// Pre-task hook spawns agents
npx claude-flow hook pre-task --description "[task]"
// Internally calls:
mcp__claude-flow__agent_spawn { type: "appropriate-agent" }

// Post-edit hook updates memory
npx claude-flow hook post-edit --file "[file]"
// Internally calls:
mcp__claude-flow__memory_usage {
  action: "store",
  key: "swarm/editor/[file]",
  namespace: "coordination",
  value: JSON.stringify({ file, changes, timestamp })
}

// Session-end hook persists state
npx claude-flow hook session-end
// Internally calls:
mcp__claude-flow__memory_persist { sessionId: "[session-id]" }
```

## Memory Coordination Protocol

All hooks follow the mandatory memory write pattern:

```javascript
// 1. STATUS - Hook starts
mcp__claude-flow__memory_usage {
  action: "store",
  key: "swarm/hooks/[hook-name]/status",
  namespace: "coordination",
  value: JSON.stringify({ status: "running", hook: "[name]" })
}

// 2. PROGRESS - Hook processes
mcp__claude-flow__memory_usage {
  action: "store",
  key: "swarm/hooks/[hook-name]/progress",
  namespace: "coordination",
  value: JSON.stringify({ progress: 50, action: "processing" })
}

// 3. COMPLETE - Hook finishes
mcp__claude-flow__memory_usage {
  action: "store",
  key: "swarm/hooks/[hook-name]/complete",
  namespace: "coordination",
  value: JSON.stringify({ status: "complete", result: "success" })
}
```

## Benefits
- ðŸ¤– Automatic agent assignment based on file type
- ðŸŽ¨ Consistent code formatting
- ðŸ§  Continuous neural pattern improvement  
- ðŸ’¾ Cross-session memory persistence via MCP tools
- ðŸ“Š Performance metrics tracking through memory
- ðŸ”„ Automatic memory coordination between agents
- ðŸŽ¯ Smart agent spawning based on task analysis

## See Also
- [Pre-Edit Hook](./pre-edit.md)
- [Post-Edit Hook](./post-edit.md)
- [Session End Hook](./session-end.md)
- [Memory Usage](../memory/memory-usage.md)
- [Agent Spawning](../agents/agent-spawning.md)