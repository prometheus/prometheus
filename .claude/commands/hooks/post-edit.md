# hook post-edit

Execute post-edit processing including formatting, validation, and memory updates.

## Usage

```bash
npx claude-flow hook post-edit [options]
```

## Options

- `--file, -f <path>` - File path that was edited
- `--auto-format` - Automatically format code (default: true)
- `--memory-key, -m <key>` - Store edit context in memory
- `--train-patterns` - Train neural patterns from edit
- `--validate-output` - Validate edited file

## Examples

### Basic post-edit hook

```bash
npx claude-flow hook post-edit --file "src/components/Button.jsx"
```

### With memory storage

```bash
npx claude-flow hook post-edit -f "api/auth.js" --memory-key "auth/login-implementation"
```

### Format and validate

```bash
npx claude-flow hook post-edit -f "config/webpack.js" --auto-format --validate-output
```

### Neural training

```bash
npx claude-flow hook post-edit -f "utils/helpers.ts" --train-patterns --memory-key "utils/refactor"
```

## Features

### Auto Formatting

- Language-specific formatters
- Prettier for JS/TS/JSON
- Black for Python
- gofmt for Go
- Maintains consistency

### Memory Storage

- Saves edit context
- Records decisions made
- Tracks implementation details
- Enables knowledge sharing

### Pattern Training

- Learns from successful edits
- Improves future suggestions
- Adapts to coding style
- Enhances coordination

### Output Validation

- Checks syntax correctness
- Runs linting rules
- Validates formatting
- Ensures quality

## Integration

This hook is automatically called by Claude Code when:

- After Edit tool completes
- Following MultiEdit operations
- During file saves
- After code generation

Manual usage in agents:

```bash
# After editing files
npx claude-flow hook post-edit --file "path/to/edited.js" --memory-key "feature/step1"
```

## Output

Returns JSON with:

```json
{
  "file": "src/components/Button.jsx",
  "formatted": true,
  "formatterUsed": "prettier",
  "lintPassed": true,
  "memorySaved": "component/button-refactor",
  "patternsTrained": 3,
  "warnings": [],
  "stats": {
    "linesChanged": 45,
    "charactersAdded": 234
  }
}
```

## See Also

- `hook pre-edit` - Pre-edit preparation
- `Edit` - File editing tool
- `memory usage` - Memory management
- `neural train` - Pattern training
