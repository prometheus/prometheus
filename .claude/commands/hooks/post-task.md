# hook post-task

Execute post-task cleanup, performance analysis, and memory storage.

## Usage

```bash
npx claude-flow hook post-task [options]
```

## Options

- `--task-id, -t <id>` - Task identifier for tracking
- `--analyze-performance` - Generate performance metrics (default: true)
- `--store-decisions` - Save task decisions to memory
- `--export-learnings` - Export neural pattern learnings
- `--generate-report` - Create task completion report

## Examples

### Basic post-task hook

```bash
npx claude-flow hook post-task --task-id "auth-implementation"
```

### With full analysis

```bash
npx claude-flow hook post-task -t "api-refactor" --analyze-performance --generate-report
```

### Memory storage

```bash
npx claude-flow hook post-task -t "bug-fix-123" --store-decisions --export-learnings
```

### Quick cleanup

```bash
npx claude-flow hook post-task -t "minor-update" --analyze-performance false
```

## Features

### Performance Analysis

- Measures execution time
- Tracks token usage
- Identifies bottlenecks
- Suggests optimizations

### Decision Storage

- Saves key decisions made
- Records implementation choices
- Stores error resolutions
- Maintains knowledge base

### Neural Learning

- Exports successful patterns
- Updates coordination models
- Improves future performance
- Trains on task outcomes

### Report Generation

- Creates completion summary
- Documents changes made
- Lists files modified
- Tracks metrics achieved

## Integration

This hook is automatically called by Claude Code when:

- Completing a task
- Switching to a new task
- Ending a work session
- After major milestones

Manual usage in agents:

```bash
# In agent coordination
npx claude-flow hook post-task --task-id "your-task-id" --analyze-performance true
```

## Output

Returns JSON with:

```json
{
  "taskId": "auth-implementation",
  "duration": 1800000,
  "tokensUsed": 45000,
  "filesModified": 12,
  "performanceScore": 0.92,
  "learningsExported": true,
  "reportPath": "/reports/task-auth-implementation.md"
}
```

## See Also

- `hook pre-task` - Pre-task setup
- `performance report` - Detailed metrics
- `memory usage` - Memory management
- `neural patterns` - Pattern analysis
