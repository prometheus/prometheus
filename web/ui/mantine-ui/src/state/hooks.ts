import { useDispatch, useSelector } from "react-redux";
import type { RootState, AppDispatch } from "./store";

// Use these typed hooks throughout the app instead of plain `useDispatch` and `useSelector`
export const useAppDispatch = useDispatch.withTypes<AppDispatch>();
export const useAppSelector = useSelector.withTypes<RootState>();
