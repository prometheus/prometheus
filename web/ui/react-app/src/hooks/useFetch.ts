import { useState, useEffect } from 'react';

export type APIResponse<T> = { status: string; data: T };

export interface FetchState<T> {
  response: APIResponse<T>;
  error?: Error;
  isLoading: boolean;
}

export const useFetch = <T extends {}>(url: string, options?: RequestInit): FetchState<T> => {
  const [response, setResponse] = useState<APIResponse<T>>({ status: 'start fetching' } as any);
  const [error, setError] = useState<Error>();
  const [isLoading, setIsLoading] = useState<boolean>(true);

  useEffect(() => {
    const fetchData = async () => {
      setIsLoading(true);
      try {
        const res = await fetch(url, { cache: 'no-store', credentials: 'same-origin', ...options });
        if (!res.ok) {
          throw new Error(res.statusText);
        }
        const json = (await res.json()) as APIResponse<T>;
        setResponse(json);
        setIsLoading(false);
      } catch (error) {
        setError(error);
      }
    };
    fetchData();
  }, [url, options]);
  return { response, error, isLoading };
};
