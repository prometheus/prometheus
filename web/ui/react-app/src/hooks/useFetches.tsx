import { useState, useEffect } from 'react';

const useFetches = <R extends any>(urls: string[], options?: RequestInit) => {
  if (!urls.length) {
    throw new Error("Doesn't have url to fetch.");
  }
  const [response, setResponse] = useState<R[]>();
  const [error, setError] = useState<Error>();

  useEffect(() => {
    const fetchData = async () => {
      try {
        const responses: R[] = await Promise.all(
          urls
            .map(async url => {
              const res = await fetch(url, options);
              if (!res.ok) {
                throw new Error(res.statusText);
              }
              const result = await res.json();
              return result.data;
            })
            .filter(Boolean) // Remove falsy values
        );
        setResponse(responses);
      } catch (error) {
        setError(error);
      }
    };
    fetchData();
  }, [urls, options]);

  return { response, error, isLoading: !response || !response.length };
};

export default useFetches;
