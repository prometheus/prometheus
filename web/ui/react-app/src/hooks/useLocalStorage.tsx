import { Dispatch, SetStateAction, useEffect, useState } from 'react';

export function useLocalStorage<S>(localStorageKey: string, initialState: S): [S, Dispatch<SetStateAction<S>>] {
  const localStorageState = JSON.parse(localStorage.getItem(localStorageKey) || JSON.stringify(initialState));
  const [value, setValue] = useState(localStorageState);

  useEffect(() => {
    const serializedState = JSON.stringify(value);
    localStorage.setItem(localStorageKey, serializedState);
  }, [localStorageKey, value]);

  return [value, setValue];
}
