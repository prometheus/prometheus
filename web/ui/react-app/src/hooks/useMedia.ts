import { useEffect, useState } from 'react';

// A hook to determine whether a CSS media query finds any matches.
const useMedia = (query: string): boolean => {
  const mediaQuery = window.matchMedia(query);
  const [matches, setMatches] = useState(mediaQuery.matches);

  useEffect(() => {
    const handler = () => setMatches(mediaQuery.matches);
    mediaQuery.addEventListener('change', handler);
    return () => mediaQuery.removeEventListener('change', handler);
  }, [mediaQuery]);

  return matches;
};

export default useMedia;
