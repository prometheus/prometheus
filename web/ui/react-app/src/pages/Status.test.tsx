import * as React from 'react';
import { shallow } from 'enzyme';
import { Status } from '.';
import { Alert } from 'reactstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import * as useFetch from '../hooks/useFetches';
import toJson from 'enzyme-to-json';

describe('Status', () => {
  afterEach(() => jest.restoreAllMocks());
  it('should render spinner while waiting data', () => {
    const wrapper = shallow(<Status />);
    expect(wrapper.find(FontAwesomeIcon)).toHaveLength(1);
  });
  it('should render Alert on error', () => {
    (useFetch as any).default = jest.fn().mockImplementation(() => ({ error: new Error('foo') }));
    const wrapper = shallow(<Status />);
    expect(wrapper.find(Alert)).toHaveLength(1);
  });
  it('should fetch proper API endpoints', () => {
    const useFetchSpy = jest.spyOn(useFetch, 'default');
    shallow(<Status pathPrefix="/path/prefix" />);
    expect(useFetchSpy).toHaveBeenCalledWith([
      '/path/prefix/api/v1/status/runtimeinfo',
      '/path/prefix/api/v1/status/buildinfo',
      '/path/prefix/api/v1/alertmanagers',
    ]);
  });
  describe('Snapshot testing', () => {
    const response = [
      {
        startTime: '2019-10-30T22:03:23.247913868+02:00',
        CWD: '/home/boyskila/Desktop/prometheus',
        reloadConfigSuccess: true,
        lastConfigTime: '2019-10-30T22:03:23+02:00',
        chunkCount: 1383,
        timeSeriesCount: 461,
        corruptionCount: 0,
        goroutineCount: 37,
        GOMAXPROCS: 4,
        GOGC: '',
        GODEBUG: '',
        storageRetention: '15d',
      },
      {
        version: '',
        revision: '',
        branch: '',
        buildUser: '',
        buildDate: '',
        goVersion: 'go1.13.3',
      },
      {
        activeAlertmanagers: [
          { url: 'https://1.2.3.4:9093/api/v1/alerts' },
          { url: 'https://1.2.3.5:9093/api/v1/alerts' },
          { url: 'https://1.2.3.6:9093/api/v1/alerts' },
          { url: 'https://1.2.3.7:9093/api/v1/alerts' },
          { url: 'https://1.2.3.8:9093/api/v1/alerts' },
          { url: 'https://1.2.3.9:9093/api/v1/alerts' },
        ],
        droppedAlertmanagers: [],
      },
    ];
    it('should match table snapshot', () => {
      (useFetch as any).default = jest.fn().mockImplementation(() => ({ response }));
      const wrapper = shallow(<Status />);
      expect(toJson(wrapper)).toMatchSnapshot();
      jest.restoreAllMocks();
    });
  });
});
