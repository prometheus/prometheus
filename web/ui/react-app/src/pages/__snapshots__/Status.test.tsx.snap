// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Status Snapshot testing should match table snapshot 1`] = `
<Fragment>
  <h2>
    Runtime Information
  </h2>
  <Table
    bordered={true}
    className="h-auto"
    responsiveTag="div"
    size="sm"
    striped={true}
    tag="table"
  >
    <tbody>
      <tr
        key="startTime"
      >
        <th
          className="capitalize-title"
          style={
            Object {
              "width": "35%",
            }
          }
        >
          Start time
        </th>
        <td
          className="text-break"
        >
          Wed, 30 Oct 2019 20:03:23 GMT
        </td>
      </tr>
      <tr
        key="CWD"
      >
        <th
          className="capitalize-title"
          style={
            Object {
              "width": "35%",
            }
          }
        >
          Working directory
        </th>
        <td
          className="text-break"
        >
          /home/boyskila/Desktop/prometheus
        </td>
      </tr>
      <tr
        key="reloadConfigSuccess"
      >
        <th
          className="capitalize-title"
          style={
            Object {
              "width": "35%",
            }
          }
        >
          Configuration reload
        </th>
        <td
          className="text-break"
        >
          Successful
        </td>
      </tr>
      <tr
        key="lastConfigTime"
      >
        <th
          className="capitalize-title"
          style={
            Object {
              "width": "35%",
            }
          }
        >
          Last successful configuration reload
        </th>
        <td
          className="text-break"
        >
          2019-10-30T22:03:23+02:00
        </td>
      </tr>
      <tr
        key="chunkCount"
      >
        <th
          className="capitalize-title"
          style={
            Object {
              "width": "35%",
            }
          }
        >
          Head chunks
        </th>
        <td
          className="text-break"
        >
          1383
        </td>
      </tr>
      <tr
        key="timeSeriesCount"
      >
        <th
          className="capitalize-title"
          style={
            Object {
              "width": "35%",
            }
          }
        >
          Head time series
        </th>
        <td
          className="text-break"
        >
          461
        </td>
      </tr>
      <tr
        key="corruptionCount"
      >
        <th
          className="capitalize-title"
          style={
            Object {
              "width": "35%",
            }
          }
        >
          WAL corruptions
        </th>
        <td
          className="text-break"
        >
          0
        </td>
      </tr>
      <tr
        key="goroutineCount"
      >
        <th
          className="capitalize-title"
          style={
            Object {
              "width": "35%",
            }
          }
        >
          Goroutines
        </th>
        <td
          className="text-break"
        >
          37
        </td>
      </tr>
      <tr
        key="GOMAXPROCS"
      >
        <th
          className="capitalize-title"
          style={
            Object {
              "width": "35%",
            }
          }
        >
          GOMAXPROCS
        </th>
        <td
          className="text-break"
        >
          4
        </td>
      </tr>
      <tr
        key="GOGC"
      >
        <th
          className="capitalize-title"
          style={
            Object {
              "width": "35%",
            }
          }
        >
          GOGC
        </th>
        <td
          className="text-break"
        />
      </tr>
      <tr
        key="GODEBUG"
      >
        <th
          className="capitalize-title"
          style={
            Object {
              "width": "35%",
            }
          }
        >
          GODEBUG
        </th>
        <td
          className="text-break"
        />
      </tr>
      <tr
        key="storageRetention"
      >
        <th
          className="capitalize-title"
          style={
            Object {
              "width": "35%",
            }
          }
        >
          Storage retention
        </th>
        <td
          className="text-break"
        >
          15d
        </td>
      </tr>
    </tbody>
  </Table>
  <h2>
    Build Information
  </h2>
  <Table
    bordered={true}
    className="h-auto"
    responsiveTag="div"
    size="sm"
    striped={true}
    tag="table"
  >
    <tbody>
      <tr
        key="version"
      >
        <th
          className="capitalize-title"
          style={
            Object {
              "width": "35%",
            }
          }
        >
          version
        </th>
        <td
          className="text-break"
        />
      </tr>
      <tr
        key="revision"
      >
        <th
          className="capitalize-title"
          style={
            Object {
              "width": "35%",
            }
          }
        >
          revision
        </th>
        <td
          className="text-break"
        />
      </tr>
      <tr
        key="branch"
      >
        <th
          className="capitalize-title"
          style={
            Object {
              "width": "35%",
            }
          }
        >
          branch
        </th>
        <td
          className="text-break"
        />
      </tr>
      <tr
        key="buildUser"
      >
        <th
          className="capitalize-title"
          style={
            Object {
              "width": "35%",
            }
          }
        >
          buildUser
        </th>
        <td
          className="text-break"
        />
      </tr>
      <tr
        key="buildDate"
      >
        <th
          className="capitalize-title"
          style={
            Object {
              "width": "35%",
            }
          }
        >
          buildDate
        </th>
        <td
          className="text-break"
        />
      </tr>
      <tr
        key="goVersion"
      >
        <th
          className="capitalize-title"
          style={
            Object {
              "width": "35%",
            }
          }
        >
          goVersion
        </th>
        <td
          className="text-break"
        >
          go1.13.3
        </td>
      </tr>
    </tbody>
  </Table>
  <h2>
    Alertmanagers
  </h2>
  <Table
    bordered={true}
    className="h-auto"
    responsiveTag="div"
    size="sm"
    striped={true}
    tag="table"
  >
    <tbody>
      <tr>
        <th>
          Endpoint
        </th>
      </tr>
      <tr
        key="https://1.2.3.4:9093/api/v1/alerts"
      >
        <td>
          <a
            href="https://1.2.3.4:9093/api/v1/alerts"
          >
            https://1.2.3.4:9093
          </a>
          /api/v1/alerts
        </td>
      </tr>
      <tr
        key="https://1.2.3.5:9093/api/v1/alerts"
      >
        <td>
          <a
            href="https://1.2.3.5:9093/api/v1/alerts"
          >
            https://1.2.3.5:9093
          </a>
          /api/v1/alerts
        </td>
      </tr>
      <tr
        key="https://1.2.3.6:9093/api/v1/alerts"
      >
        <td>
          <a
            href="https://1.2.3.6:9093/api/v1/alerts"
          >
            https://1.2.3.6:9093
          </a>
          /api/v1/alerts
        </td>
      </tr>
      <tr
        key="https://1.2.3.7:9093/api/v1/alerts"
      >
        <td>
          <a
            href="https://1.2.3.7:9093/api/v1/alerts"
          >
            https://1.2.3.7:9093
          </a>
          /api/v1/alerts
        </td>
      </tr>
      <tr
        key="https://1.2.3.8:9093/api/v1/alerts"
      >
        <td>
          <a
            href="https://1.2.3.8:9093/api/v1/alerts"
          >
            https://1.2.3.8:9093
          </a>
          /api/v1/alerts
        </td>
      </tr>
      <tr
        key="https://1.2.3.9:9093/api/v1/alerts"
      >
        <td>
          <a
            href="https://1.2.3.9:9093/api/v1/alerts"
          >
            https://1.2.3.9:9093
          </a>
          /api/v1/alerts
        </td>
      </tr>
    </tbody>
  </Table>
</Fragment>
`;
